enhancements/ingress/lb-allowed-source-ranges.md

```diff
@@ -0,0 +1,261 @@
+- On OpenShift 4.12, if `service.beta.kubernetes.io/load-balancer-source-ranges` annotation is set.
+
+### API Extensions
+
+This proposal will modify the `IngressController` API by adding a new
+field called `AllowedSourceRanges` to the `LoadBalancerStrategy` 
+struct type and a new type `CIDR` that is a string in CIDR notation.
+
+[OpenShift API Changes](https://github.com/openshift/api/pull/1222) contains the literal API changes that we're
+proposing for this EP.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The value set for `AllowedSourceRanges` will be used to set `.Spec.LoadBalancerSourceRanges` 
+of the LoadBalancer Service in front of the router in cloud environments.
+
+### Risks and Mitigations
+
+One risk of this proposal is that customers may have already set the
+`.Spec.LoadBalancerSourceRanges` field of the LoadBalancer Service in a previous version. 
+This risk is mitigated by marking the cluster as non-upgradable in such case.
+
+### Drawbacks
+
+This enhancement may bring additional complexity for the customers who are 
+used to modify directly the `Spec.LoadBalancerSourceRanges` of the LoadBalancer 
+Service for the same purposes. 
+
+## Design Details
+
+### Open Questions [optional]
+
+1. On Kubernetes API reference, the field 
+[`loadBalancerSourceRanges`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#service-v1-core) 
+has the warning "This field will be ignored if the cloud-provider does not support the feature.".
+However, there is no information about which platforms support this. Is there any platform
+OpenShift runs on which does not support this feature?
+
+A. I could only find this statement from Kubernetes
+[v1.18 website repository](https://github.com/kubernetes/website/blob/dev-1.18/content/en/docs/tasks/access-application-cluster/configure-cloud-provider-firewall.md): 
+"This feature is currently supported on Google Compute Engine, Google Kubernetes Engine, 
+AWS Elastic Kubernetes Service, Azure Kubernetes Service, and IBM Cloud Kubernetes Service". 
+The current version of the document does not mention `.Spec.LoadBalancerSourceRanges` anymore.
+
+### Test Plan
+
+The operator that manages the IngressController Deployment and related
+resources has unit test coverage; for this enhancement, the unit tests are
+expanded to cover the additional functionality.  
+
+The operator has end-to-end tests; for this enhancement, a test is added that verifies
+the following scenarios:
+
+1. Adding/deleting/changing `AllowedSourceRanges` on the IngressController and observing `.Spec.LoadBalancerSourceRanges` on the LoadBalancer Service.
+
+2. Adding `service.beta.kubernetes.io/load-balancer-source-ranges` annotation to a LoadBalancer Service and observing cluster's upgradable status.
+
+There is an upgrade test in openshift/origin to make sure that the migration workflow explained under **Variation** above works seamlessly.
```

Comment:
```suggestion
This enhancement adds an upgrade test in openshift/origin to make sure that the migration workflow explained under **Variation** above works seamlessly.
```
Please provide a little more detail about the test.  For example, you could reference the `"k8s.io/kubernetes/test/e2e/upgrades".Test` interface and outline the steps in terms of it; I have the following in mind:

1. In `Setup`, create an IngressController that requests an LB, and set the `service.beta.kubernetes.io/load-balancer-source-ranges` annotation on the resulting LoadBalancer-type service.
2. In `Test`, periodically check the service to verify that nothing updates the annotation or the `Spec.LoadBalancerSourceRanges` field on the service during the upgrade.
3. In `Test`, verify again when the upgrade is done that nothing has updated the annotation or `Spec.LoadBalancerSourceRanges`, and verify that the operator has set `Upgradeable=False`.
4. Then set `Spec.LoadBalancerSourceRanges`, then remove the annotation, and finally verify that the operator has set `Upgradeable=True`.

If necessary, check with someone more versed in upgrade tests to make sure this outline, or whatever approach you had in mind if it differs from what I've described, makes sense to do in an upgrade test and won't break anything (in particular, I'm not certain that causing an operator to set `Upgradeable=False` is acceptable in an upgrade test).